shared_code/base_scraper.py

```diff
@@ -138,6 +140,84 @@ def is_new_data_available(self, client: MongoClient) -> bool:
+            foundBadData = False
+            cityMissing = False
+            stateMissing = False
+            zipMissing = False
+
+            #Check if the row called 'name' is not null or 'NONE'
+            if nullValues['name'][index]:
+                logger.info(" row " + str(index + 1) + ": \'name\' cannot be a null value")
+                foundBadData = True
+            elif row['name'].upper() == "NONE":
+                logger.info(" row " + str(index + 1) + ": \'name\' cannot be \'" + row['name'] + "\'")
+                foundBadData = True
+
+            #Check if the row called 'city' is not null or 'NONE'
+            if nullValues['city'][index]:
+                logger.info(" row " + str(index + 1) + ": \'city\' cannot be a null value")
+                cityMissing = True
+            elif row['city'].upper() == 'NONE':
+                logger.info(" row " + str(index + 1) + ": \'city\' cannot be \'" + row['city'] + "\'")
+                cityMissing = True
+
+            # Check if the row called 'state' is not null or 'NONE'
+            if nullValues['state'][index]:
+                logger.info(" row " + str(index + 1) + ": \'state\' cannot be a null value")
+                stateMissing = True
+            elif row['state'].upper() == 'NONE':
+                logger.info(" row " + str(index + 1) + ": \'state\' cannot be \'" + row['state'] + "\'")
+                stateMissing = True
+
+            # Check if the row called 'zip' is not null or 'NONE'
+            if nullValues['zip'][index]:
+                logger.info(" row " + str(index + 1) + ": \'zip\' cannot be a null value")
+                zipMissing = True
+            elif str(row['zip']).upper() == 'NONE':
+                logger.info(" row " + str(index + 1) + ": \'zip\' cannot be \'" + row['zip'] + "\'")
+                zipMissing = True
+
+            # Check if the row called 'serviceSummary' is not null or 'NONE'
+            if nullValues['serviceSummary'][index]:
+                logger.info(" row " + str(index + 1) + ": \'serviceSummary\' was a null value but must be one of the " +
+                            "following: " + str(goodSummaries))
+                foundBadData = True
+            elif row['serviceSummary'] not in goodSummaries:
+                logger.info(" row " + str(index + 1) + ": \'serviceSummary\' got " + row['serviceSummary'] +
+                            " but must be one of the following: " + str(goodSummaries))
+                foundBadData = True
+
+            #The row should be removed if at least one of the following is true:
+                #The 'name' column is missing
+                #The 'city', 'state', and 'zip' columns are all either null or 'NONE'
+                #The 'serviceSummary' column is missing
+            if cityMissing and stateMissing and zipMissing:
+                foundBadData = True
+            if foundBadData:
+                df.drop(index, inplace=True)
```

Comment:
In order to prevent the insertion of bad data, I decided to write the `validate_data` function that would remove any rows with bad data. The data frame's rows still keep their original indices. However, this means that the indices of the data frame's rows might not form a list of consecutive nonnegative numbers starting with 0. To enable my code to consider any change in the order of indices, I had to replace the line `for i in tqdm(range(len(df))):` with `for i, row in tqdm(df.iterrows()):`.

I did extensive testing using 20 cases in the CSV file, and it worked as expected.